File: 2024/02/02.js
const {
  readFileSync,
  toTally,
  join,
  resolve,
  transpose,
} = require("../../js_modules/helpers.js");

function filepath(filename, subfolder = "") {
  return join(resolve(__dirname), subfolder, filename);
}

let path = filepath("input.txt");
// path = filepath("exinput.txt");
let lines = readFileSync(path, "utf-8").trim().split("\n");

// Part 1

let reports = lines.map((line) => line.match(/\d+/g).map(Number));

function isSafe(nums) {
  let deltas = nums.slice(0, nums.length - 1).map((x, i) => nums[i + 1] - x);
  let max = Math.max(...deltas);
  let min = Math.min(...deltas);
  return Math.max(...[max, min].map(Math.abs)) <= 3 && max * min > 0;
}

function count(iter, condition) {
  return iter.reduce((count, x) => count + (condition(x) ? 1 : 0), 0);
}

console.log(count(reports, isSafe));
// Part 2

function isSafeDampened(nums) {
  return nums.some((x, i) =>
    isSafe(nums.slice(0, i).concat(nums.slice(i + 1))),
  );
}

console.log(count(reports, isSafeDampened));

File: 2024/02/02.py
path = r"input.txt"
with open(path, "r") as f:
    reports = [[int(x) for x in line.strip().split()] for line in f.readlines()]


def is_safe(report):
    deltas = [report[i + 1] - report[i] for i in range(len(report) - 1)]
    return (
        len(set(max(-1, min(1, x)) for x in deltas)) == 1
        and max(abs(x) for x in deltas) < 4
    )


part1 = sum(1 for r in reports if is_safe(r))
print(part1)

part2 = sum(
    1 for r in reports if any([is_safe(r[0:i] + r[i + 1 :]) for i in range(len(r))])
)
print(part2)
